core/src/main/java/org/apache/calcite/adapter/enumerable/EnumTraitsUtils.java

```diff
@@ -148,4 +149,52 @@ private static boolean isCollationOnTrivialExpr(
       return null;
     }
   }
+
+  // EnumerableNestedLoopJoin and EnumerableBatchNestedLoopJoin traits passdown shall only
+  // pass through collation to left input. It is because for EnumerableNestedLoopJoin always
+  // uses left input as the outer loop, thus only left input can preserve ordering.
+  // Push sort both to left and right inputs does not help right outer join. It's because in
+  // implementation, EnumerableNestedLoopJoin produces (null, right_unmatched) all together,
+  // which does not preserve ordering from right side.
+  static Pair<RelTraitSet, List<RelTraitSet>> passThroughTraitsForNestedLoopJoin(
```

Comment:
HashJoin and Correlate can reuse it too.

Reply:
Nice catch. Will migrate them too to this PR.

Reply:
The class name is misleading. Should be EnumerableTraitUtil, Enum is a different thing.

Reply:
Pushed a commit to make HashJoin and Correlate reuse the code. Also rename this class to `EnumerableTraitUtils`